### DISCUSSION TITLE
[Proposal] Automatically inject the user in an entity

### DISCUSSION THREAD

**Discussion by kevin-verschaeve** [symfony/symfony#44562] [Q&A]
### Description

Hello,
I was wondering if having a solution to automatically inject the current logged in user into an entiy on save would be a good idea.

In my project, I have several entities for which I need to store the current user in a property, so we use entity listener.
As I think this is a common case, I think it could be done by Symfony.

I made a POC in FrameworkBundle to try this, and my current solution is to create an EntityListener + a new custom php attribute.

Here is the usage for a user : 
```
/**
 * @ORM\Entity
 * @ORM\EntityListeners({"Symfony\Bundle\FrameworkBundle\EventListener\InjectSecurityUserListener"})
 */
class Example
{
    ...

    /**
     * @ORM\ManyToOne(targetEntity=User::class)
     */
    #[InjectSecurityUser]
    private $user;
```

I don't know if this is the best solution, I'm open to suggestions if you have any.
I can make a pull request if you think this idea could be nice to have.

Thanks for your feedback !

**Comment by GromNaN** [comment]
Using a Doctrine EventListener is a good first step. To detect entities that needs this behavior, I would use an interface implemented only by involved entity class.

This is how the Blamable behavior works. https://github.com/KnpLabs/DoctrineBehaviors/blob/v2.0.8/docs/blameable.md

**Reply by kevin-verschaeve** [reply]
using an interface is a good idea indeed, but it does not allow to inject the user in any property and it allows to inject in one property only on the class (I don't have the use case but that could be something wanted).
Maybe a mix of the interface, and the annotation would be great to let the developer inject the user on the property(ies) he want ?

**Reply by GromNaN** [reply]
The method to inject the user don't have to be the standard setter of the field. 

```php
interface InjectableUserInterface {
    public function injectUser(User $user);
}

class Example implements InjectableUserInterface {
    public function injectUser(User $user) {
        $this->user = $user;
    }
}
```
